### PR TITLE
Teams mobile - invite contacts

### PR DESCRIPTION
**Pull request by buoyad**
- [x] Add native SMS helper
- [x] Add dep to get contacts
- [x] Requesting contact permissions flow
- [x] Permissions failed screen (w/ instructions on granting them)
- [x] Display contacts UI (preliminary)
- [x] Plumb through RPC for retrieving token for SMS invites (placeholder)
- [x] Plumb through RPC for sending out email invites
- [x] Split container for desktop & mobile (mobile w/ one-at-a-time invites, desktop aggregate)
- [x] Map pending invites to control `Invited!` contacts on UI
- [x] Add loading map for invites in flight
- [x] Add filter for contacts
- [x] Hoist contact processing to container
- [x] Finalize UI

### CODE REVIEW

**Comment by buoyad** [comment]
Currently vendors a fork of `react-get-contacts` under `react-native/` until we decide how to handle that

**Comment by keybase-ci-visdiff** [comment]
:no_entry: Error rendering the commits 34b455fceb98...4c41d79fff2e on linux. <details><summary>:mag_right: 22 errored, 7 new, 13 changed</summary>

 * Error in 4c41d79fff2e: **Chat_Inbox-Selected_Normal-ERROR**  
   Chat_Inbox-Selected_Normal-ERROR rendered
 * Error in 4c41d79fff2e: **Chat_Inbox-Selected_Muted-ERROR**  
   Chat_Inbox-Selected_Muted-ERROR rendered
 * Error in 34b455fceb98: **Devices_Devices_List-Devices-ERROR**  
   Devices_Devices_List-Devices-ERROR rendered
 * Error in 4c41d79fff2e: **Devices_Devices_List-Devices-ERROR**  
   Devices_Devices_List-Devices-ERROR rendered
 * Error in 34b455fceb98: **Chat_List-Empty-ERROR**  
   Chat_List-Empty-ERROR rendered
 * Error in 4c41d79fff2e: **Chat_List-Empty-ERROR**  
   Chat_List-Empty-ERROR rendered
 * Error in 34b455fceb98: **Chat_Input-Normal-ERROR**  
   Chat_Input-Normal-ERROR rendered
 * Error in 4c41d79fff2e: **Chat_Input-Normal-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Input-Normal-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-You_Rekey_Selected-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-You_Rekey_Selected-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-You_Rekey_Selected-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-You_Rekey_Selected-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-You_Rekey-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-You_Rekey-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-You_Rekey-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-You_Rekey-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-Part_Rekey_Selected-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Part_Rekey_Selected-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-Part_Rekey_Selected-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Part_Rekey_Selected-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-Part_Rekey-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Part_Rekey-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-Part_Rekey-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Part_Rekey-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-Long_Top-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Long_Top-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-Long_Top-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Long_Top-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-Long_Bottom-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Long_Bottom-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-Long_Bottom-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Long_Bottom-ERROR-4c41d79fff2e.png)
 * Error in 34b455fceb98: **Chat_Inbox-Empty-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Empty-ERROR-34b455fceb98.png)
 * Error in 4c41d79fff2e: **Chat_Inbox-Empty-ERROR** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Empty-ERROR-4c41d79fff2e.png)
 * Added: **Chat_Inbox-Normal** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Inbox-Normal-4c41d79fff2e.png)
 * Added: **Chat_Banner_Invite-Invite** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Banner_Invite-Invite-4c41d79fff2e.png)
 * Added: **Chat_Banner_Info-Info** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Banner_Info-Info-4c41d79fff2e.png)
 * Added: **Chat_Banner_Error-Error** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Banner_Error-Error-4c41d79fff2e.png)
 * Added: **Chat_Banner_Broken-Broken_Tracker_3** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Banner_Broken-Broken_Tracker_3-4c41d79fff2e.png)
 * Added: **Chat_Banner_Broken-Broken_Tracker_2** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Banner_Broken-Broken_Tracker_2-4c41d79fff2e.png)
 * Added: **Chat_Banner_Broken-Broken_Tracker_1** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Chat_Banner_Broken-Broken_Tracker_1-4c41d79fff2e.png)
 * Changed: **Text-Terminal_Empty_Information** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Terminal_Empty_Information-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Terminal_Empty_Information-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Terminal_Empty_Information.png)
 * Changed: **Text-Terminal_Empty_Announcements** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Terminal_Empty_Announcements-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Terminal_Empty_Announcements-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Terminal_Empty_Announcements.png)
 * Changed: **Text-Header_Big_Success** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Header_Big_Success-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Header_Big_Success-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Header_Big_Success.png)
 * Changed: **Text-Body_Wrap_Announcements** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Wrap_Announcements-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Wrap_Announcements-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Wrap_Announcements.png)
 * Changed: **Text-Body_Small_Semibold_Italic_Documentation** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Small_Semibold_Italic_Documentation-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Small_Semibold_Italic_Documentation-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Small_Semibold_Italic_Documentation.png)
 * Changed: **Text-Body_Small_Italic_High_Risk** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Small_Italic_High_Risk-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Small_Italic_High_Risk-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Small_Italic_High_Risk.png)
 * Changed: **Text-Body_Secondary_Link_High_Risk** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Secondary_Link_High_Risk-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Secondary_Link_High_Risk-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Text-Body_Secondary_Link_High_Risk.png)
 * Changed: **Colors-white_20_rgba_255_255_255_0_20** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-white_20_rgba_255_255_255_0_20-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-white_20_rgba_255_255_255_0_20-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-white_20_rgba_255_255_255_0_20.png)
 * Changed: **Colors-red_75_rgba_255_0_0_0_75** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-red_75_rgba_255_0_0_0_75-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-red_75_rgba_255_0_0_0_75-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-red_75_rgba_255_0_0_0_75.png)
 * Changed: **Colors-orange_ff_6_f_21** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-orange_ff_6_f_21-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-orange_ff_6_f_21-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Colors-orange_ff_6_f_21.png)
 * Changed: **Buttons-Follow_small_waiting** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Follow_small_waiting-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Follow_small_waiting-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Follow_small_waiting.png)
 * Changed: **Buttons-Follow_small** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Follow_small-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Follow_small-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Follow_small.png)
 * Changed: **Buttons-Danger_small** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Danger_small-34b455fceb98.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Danger_small-4c41d79fff2e.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1508525523984-34b455fceb98-4c41d79fff2e-linux/Buttons-Danger_small.png)
</summary>

**Comment by buoyad** [comment]
@keybase/react-hackers ready! 

This had more moving parts than I've worked with so far, so please fire away at the way this is structured. Also dropped in a placeholder RPC for inviting via seitan token.

File: shared/package.json [hunk 1]
```diff
@@ -112,6 +112,7 @@
     "react-list": "0.8.8",
     "react-native": "0.49.3",
     "react-native-camera": "0.10.0",
+    "react-native-contacts": "^1.0.2",
```
Comment (chrisnojima) [inline shared/package.json]: we use exact semvers

File: shared/react-native/android/app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -12,7 +12,8 @@
     <uses-permission android:name="${applicationId}.permission.C2D_MESSAGE" />
     <uses-permission android:name="android.permission.VIBRATE" />
     <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
-
+    <uses-permission android:name="android.permission.READ_CONTACTS" />
+    <uses-permission android:name="android.permission.READ_PROFILE" />
```
Comment (chrisnojima) [inline shared/react-native/android/app/src/main/AndroidManifest.xml]: do we actually need read_profile?

File: shared/react-native/react-native-contacts/.gitignore [hunk 3]
```diff
@@ -0,0 +1,8 @@
+node_modules
```
Comment (chrisnojima) [inline shared/react-native/react-native-contacts/.gitignore]: lets use their module in our package.json and not copy the source into our repo

File: shared/teams/invite-by-email/container.native.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      title: 'Invite contacts',
+      headerStyle: {borderBottomWidth: 0},
+    })),
+    withState('contacts', '_setContacts', []),
+    withState('hasPermission', '_setHasPermission', true),
+    lifecycle({
+      componentWillMount() {
+        // TODO test this permission flow on a real build
+        if (isAndroid) {
+          Contacts.requestPermission((_, granted) => {
```
Comment (chrisnojima) [inline shared/teams/invite-by-email/container.native.js]: maybe we should pull these out into little helpers so its not all inline in these big blocks

File: shared/teams/invite-by-email/index.native.js [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        To fix this, please open Settings > Keybase and check off 'Allow Keybase to access Contacts'.
+      </Text>
+      <Button type="Primary" label="Open settings" onClick={openSettings} />
+    </Box>
+  </Box>
+)
+
+const contactRow = (i: number, props: ContactRowProps) => {
+  const source = typeof props.contact.thumbnailPath === 'string'
+    ? {uri: 'file://' + props.contact.thumbnailPath}
```
Comment (chrisnojima) [inline shared/teams/invite-by-email/index.native.js]: prefer using `file://${props.contact.thumbnailPath}`

File: shared/teams/invite-by-email/index.native.js [hunk 6]
```diff
@@ -0,0 +1,230 @@
+// @flow
+import * as React from 'react'
+import {Avatar, Box, Button, ClickableBox, Icon, Input, List, Text} from '../../common-adapters'
+import {globalStyles, globalMargins, globalColors} from '../../styles'
+import {NativeImage} from '../../common-adapters/native-wrappers.native'
```
Comment (chrisnojima) [inline shared/teams/invite-by-email/index.native.js]: this is exposed if you import common-adapters/index.native, use that instead and add it to the other imports you have above

File: shared/teams/invite-by-email/index.native.js [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+type State = {
+  filter: string,
+}
+
+class InviteByEmail extends React.Component<MobileProps, State> {
+  constructor(props: MobileProps) {
+    super(props)
+    this.state = {
```
Comment (chrisnojima) [inline shared/teams/invite-by-email/index.native.js]: you can remove this constructor and instead define state as a class property just do
```
class InviteByEmail extends... {
  state = {
    filter: ''
  }
}
```

File: shared/teams/invite-by-email/index.native.js [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  _onSelectContact(contact: ContactDisplayProps) {
+    if (this._isSelected(contact.recordID)) {
+      this._removeInvitee(contact)
+    } else {
+      this._addInvitee(contact)
+    }
+  }
+
+  _isSelected(id: string): boolean {
+    return this.props.invited.findIndex(rec => rec.contactID === id) >= 0
```
Comment (chrisnojima) [inline shared/teams/invite-by-email/index.native.js]: can be 
```
!!this.props.invited.find(rec => rec.contactID === id)
```

File: shared/teams/invite-by-email/index.native.js [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    }
+  }
+
+  _trim(s: ?string): string {
+    return (s && s.replace(/^[^a-z0-9@.]/i, '').toLowerCase()) || ''
+  }
+
+  render() {
+    // Filter before adding props to avoid a long map fcn
+    const filteredContactRows = this.props.contactRowProps.filter(contact => {
```
Comment (chrisnojima) [inline shared/teams/invite-by-email/index.native.js]: i don't understand why we have complicated stuff here and also in the container, can't this move into the container or vice versa

File: shared/react-native/android/app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (buoyad) [reply shared/react-native/android/app/src/main/AndroidManifest.xml]: Hmm I tested again without it, and didn't get the crasher. Maybe it was unrelated, I removed the permission.

**Pull request merged**